deepmd/pt_expt/train/training.py

```
@@ -947,6 +1125,35 @@ def _compile_model(self, compile_opts: dict[str, Any]) -> None:
+        # Pre-pass: group tasks by structure key and auto-detect per-task buffers.
+        # Grouping is needed so _detect_task_buffers can diff buffer identities
+        # across all tasks that share the same compiled graph.
+        _key_for: dict[str, tuple[int, ...]] = {}
+        _groups: defaultdict[tuple[int, ...], list[str]] = defaultdict(list)
+        for task_key in self.model_keys:
+            sk = _get_model_structure_key(wrapper_mod.model[task_key])
+            _key_for[task_key] = sk
+            _groups[sk].append(task_key)
+
+        _task_bufs_for: dict[str, dict[str, torch.Tensor]] = {}
+        for sk, group_keys in _groups.items():
```

Ruff (0.15.14)

[warning] Loop control variable `sk` not used within loop body

(B007)